### PR TITLE
Initial checkin of Notification service

### PR DESCRIPTION
**Pull request by rynowak**
This implementation uses Reflection.Emit to generate proxies for
duck-typing of event data on net45 and dnxcore50.

For now, the code that does a 'splat' of the top-level event data is still
using reflection instead of code-generation. The next update will address
this and add more test coverage for this part of the code.

### CODE REVIEW

File: src/Microsoft.Framework.Notification/Internal/Converter.cs [hunk 1]
```diff
@@ -0,0 +1,372 @@
...
+using System.Diagnostics;
+using System.Linq;
+using System.Reflection;
+using System.Reflection.Emit;
+
+namespace Microsoft.Framework.Notification.Internal
+{
+    public class Converter
+    {
+        private static int _counter = 0;
```
Comment (Eilon) [inline src/Microsoft.Framework.Notification/Internal/Converter.cs]: Don't initialize me. Plus recommend rename: `_proxyCounter`?

File: src/Microsoft.Framework.Notification/Resources.resx [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+    <value>2.0</value>
+  </resheader>
+  <resheader name="reader">
+    <value>System.Resources.ResXResourceReader, System.Windows.Forms, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089</value>
+  </resheader>
+  <resheader name="writer">
+    <value>System.Resources.ResXResourceWriter, System.Windows.Forms, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089</value>
+  </resheader>
+  <data name="Converter_PropertyMustHaveGetter" xml:space="preserve">
+    <value>The property '{0}' on type '{1}' must define a getter to support proxy generation. </value>
```
Comment (Eilon) [inline src/Microsoft.Framework.Notification/Resources.resx]: Trailing space :trollface:

**Comment by Eilon** [comment]
Overall looks good to me. I didn't look at the codegen stuff too closely, though. I have niggling concerns about the possibility of bugs in there around the precision of how it looks at all the different kinds of properties. I didn't spot anything specifically wrong looking, but it's just so easy to get that wrong. E.g. property accessibility, weird generic stuff, nested types, explicit interface implementations, indexers, mixed accessibility of getter/setter/property, blah blah blah.

File: src/Microsoft.Framework.Notification/Internal/Converter.cs (hunk 1 above)
Comment (rynowak) [inline src/Microsoft.Framework.Notification/Internal/Converter.cs]: :+1:

File: src/Microsoft.Framework.Notification/Resources.resx (hunk 2 above)
Comment (rynowak) [inline src/Microsoft.Framework.Notification/Resources.resx]: :fireworks:

**Comment by rynowak** [comment]
> property accessibility, weird generic stuff, nested types, explicit interface implementations, indexers, mixed accessibility of getter/setter/property, blah blah blah

Will add a few tests. At this point the effect of trying one of these cases would be to find out if the error message is at least reasonable. We're not trying to support any complicated scenarios through the proxy, just to try and build something reasonably useful.

**Comment by rynowak** [comment]
Actually, let me walk it back a sec.

> E.g. property accessibility, weird generic stuff, nested types, explicit interface implementations, indexers, mixed accessibility of getter/setter/property, blah blah blah.

All of these tests need to be done based on the declaration of the property done on the **interface**. So that pretty much leaves indexers as the only special case right?

**Comment by Eilon** [comment]
Ah, good point about the interface. Ok then derived interfaces might be tricky :smile: Gimme at least something!!! :smile:

**Pull request closed without merging**